File: 백트래킹/boj_9663_N-Queen.c
//https://www.acmicpc.net/problem/9663
#include <stdlib.h>
#include <stdio.h>

int answer = 0;
bool check(int *graph, int row) {
	for (int i = row-1; i >= 0; i--) {
		//위의 행들이 세로로 있거나 대각에 있으면 return false
		if (graph[i] == graph[row]) {
			return false;
		}
		if (abs(graph[row] - graph[i]) == abs(row - i)) {
			return false;
		}
	}
	return true;
}


void backTracking(int *graph,int row, int n) {

	if (row == n) {
		answer++;
		return;
	}
	
	//체스판의 각 행마다 모두 퀸을 놓아보고, 되면 다음 행, 안 되면 다음 열
	for (int col = 0; col < n; col++) {
		graph[row] = col;
		if (check(graph, row)) {
			backTracking(graph, row + 1, n);
		}
		graph[row] = 0;
	}
	
}

int main() {

	int n;
	scanf("%d", &n);

	//동적 배열
	int *graph = (int*)malloc(sizeof(int) * n);

	for (int i = 0; i < n; i++) {
		graph[i] = 0;
	}
	backTracking(graph,0, n);
	printf("%d", answer);
	return 0;
}

File: 백트래킹/boj_9663_N-Queen.kt
//https://www.acmicpc.net/problem/9663
import kotlin.math.*

var answer=0

fun isPossible(row : Int,board : IntArray) : Boolean{

    for(i in 0 until row){
    //현재 둘 퀸의 자리가 이전까지 퀸들의 행과열,대각선 자리라면 둘 수 없음
        if(board[i]==board[row] || (row-i == abs(board[row]-board[i])))
            return false
    }
    return true
}

fun dfs(row : Int,board : IntArray,n : Int){

	//한 행씩 퀸을 두면서 내려왔으므로, 현재 행이 n과 같아진다면 n개의 퀸을 두었다는 의미
    if(row == n){
        answer++
        return
    }
    for(i in 0 until n ){
        board[row] = i
        //현재 행의 열에 퀸을 둘 수 있으면 다음 행 이어서 탐색
        //현재 행의 열에 퀸을 둘 수 없다면 현재 행의 다음 열 이어서 탐색
        if(isPossible(row,board)){
            dfs(row+1,board,n)
        }
    }

}

fun main() = with(System.out.bufferedWriter()){
    //1<=n<=15
    val br =System.`in`.bufferedReader()
    val n = Integer.parseInt(br.readLine())
    val board = IntArray(n)
    dfs(0,board,n)
    write("$answer")
    close()
}
